File: NodeJSWebapp/mssql.js
const sql = require('mssql');

const dbConfig = {
  user: 'sa',
  password: 'YourStrong!Passw0rd',
  server: 'db', // o IP o hostname
  database: 'web_app',
  options: {
    encrypt: true, // si usas Azure, etc.
    trustServerCertificate: true // para dev local sin certificado
  }
};

async function connectDb() {
  try {
    const pool = await sql.connect(dbConfig);
    console.log('Conectado a MSSQL');
    return pool;
  } catch (error) {
    console.error('Error de conexión MSSQL:', error);
    return null;
  }
}

// queryDb para MSSQL
async function queryDb(query, params = []) {
  const pool = await connectDb();
  if (!pool) return null;

  try {
    const request = pool.request();
    params.forEach((param, i) => {
      // SQL Server usa parámetros nombrados @p1, @p2 ...
      request.input(`p${i + 1}`, param);
    });
    // Reemplazar ? por @p1, @p2 ...
    const namedQuery = query.replace(/\?/g, (_, i) => `@p${i + 1}`);
    const result = await request.query(namedQuery);
    return result.recordset;
  } catch (error) {
    console.error('Error en consulta MSSQL:', error);
    return null;
  } finally {
    await pool.close();
    console.log('Conexión MSSQL cerrada');
  }
}

// Ejemplo funciones
async function insertUser(username, firstname, lastname, email, password, avatar) {
  const query = `
    INSERT INTO users (username, firstname, lastname, email, password, avatar)
    VALUES (?, ?, ?, ?, ?, ?)
  `;
  const params = [username, firstname, lastname, email, password, avatar];
  return await queryDb(query, params);
}

async function getUserByUsername(username) {
  const query = `SELECT * FROM users WHERE username = ?`;
  return await queryDb(query, [username]);
}

module.exports = {
  queryDb,
  insertUser,
  getUserByUsername
};

File: PythonWebApp/mssql.py
import pyodbc

def connect_db():
    try:
        connection = pyodbc.connect(
            'DRIVER={ODBC Driver 17 for SQL Server};'
            'SERVER=db;DATABASE=web_app;UID=sa;PWD=YourStrong!Passw0rd'
        )
        print('Conectado a la base de datos (SQL Server)')
        return connection
    except Exception as e:
        print(f'Error de conexión: {e}')
        return None

def close_connection(connection):
    connection.close()
    print('Conexión cerrada')

def query_db(query, params=None):
    connection = connect_db()
    if connection:
        cursor = connection.cursor()
        try:
            cursor.execute(query, params or ())
            if query.strip().lower().startswith("select"):
                columns = [column[0] for column in cursor.description]
                result = [dict(zip(columns, row)) for row in cursor.fetchall()]
            else:
                result = {'affected_rows': cursor.rowcount}
            connection.commit()
            return result
        except Exception as e:
            print(f'Error en la consulta: {e}')
            return None
        finally:
            cursor.close()
            close_connection(connection)

def insert_user(username, firstname, lastname, email, password, avatar):
    query = """
        INSERT INTO users (username, firstname, lastname, email, password, avatar)
        VALUES (?, ?, ?, ?, ?, ?)
    """
    params = (username, firstname, lastname, email, password, avatar)
    return query_db(query, params)

def get_user_by_username(username):
    query = "SELECT * FROM users WHERE username = ?"
    return query_db(query, (username,))
